File: tests/static-interval-tree.js
'use strict';

const forEach = require('obliterator/foreach');
const protocol = process.env.MNEMONIST_TRANSPORT === 'protocol'
  ? require('../adapter/_protocol.js')
  : null;

function StaticIntervalTree(intervals, getters) {
  this.intervals = intervals;
  this.size = intervals.length;
  this._inner = protocol && !getters ? protocol.create('static-interval-tree', [intervals]) : null;
  if (this._inner) {
    this.height = this._inner.call('height', []);
    return;
  }
  this._start = getters ? getters[0] : interval => interval[0];
  this._end = getters ? getters[1] : interval => interval[1];
  this.height = Math.ceil(Math.log2(this.size + 1));
  this._tree = new Array(Math.max(0, Math.pow(2, this.height) - 1)).fill(-1);
  this._maxEnds = new Array(this.size);
  const indices = intervals.map((_, index) => index).sort((a, b) => this._start(intervals[a]) - this._start(intervals[b]));

  const build = (node, low, high) => {
    const mid = Math.floor(low + (high - low) / 2);
    const index = indices[mid];
    this._tree[node] = index;
    let maxIndex = index;
    if (low < mid) {
      const left = build(node * 2 + 1, low, mid - 1);
      if (this._end(intervals[left]) > this._end(intervals[maxIndex])) maxIndex = left;
    }
    if (mid < high) {
      const right = build(node * 2 + 2, mid + 1, high);
      if (this._end(intervals[right]) > this._end(intervals[maxIndex])) maxIndex = right;
    }
    this._maxEnds[index] = maxIndex;
    return maxIndex;
  };

  if (this.size) build(0, 0, this.size - 1);
}

StaticIntervalTree.prototype._query = function(start, end) {
  if (this._inner) {
    return start === end ?
      this._inner.call('queryPoint', [start]) :
      this._inner.call('queryInterval', [start, end]);
  }
  const matches = [];
  if (!this.size) return matches;
  const stack = [0];
  while (stack.length) {
    const node = stack.pop();
    const index = this._tree[node];
    if (index < 0) continue;
    if (start > this._end(this.intervals[this._maxEnds[index]])) continue;
    const left = node * 2 + 1;
    if (left < this._tree.length && this._tree[left] >= 0) stack.push(left);
    const interval = this.intervals[index];
    const intervalStart = this._start(interval);
    const intervalEnd = this._end(interval);
    if (end >= intervalStart && start <= intervalEnd) matches.push(interval);
    if (end < intervalStart) continue;
    const right = node * 2 + 2;
    if (right < this._tree.length && this._tree[right] >= 0) stack.push(right);
  }
  return matches;
};

StaticIntervalTree.prototype.intervalsContainingPoint = function(point) {
  return this._query(point, point);
};

StaticIntervalTree.prototype.intervalsOverlappingInterval = function(interval) {
  if (this._inner) return this._query(interval[0], interval[1]);
  return this._query(this._start(interval), this._end(interval));
};

StaticIntervalTree.from = function(iterable, getters) {
  const intervals = [];
  forEach(iterable, (value, key) => {
    intervals.push(Array.isArray(value) || getters ? value : [key, value]);
  });
  return new StaticIntervalTree(intervals, getters);
};

module.exports = StaticIntervalTree;

File: src/static_interval_tree.rs
use serde_json::Value;

#[derive(Debug, Clone, PartialEq)]
pub struct Interval {
    pub start: f64,
    pub end: f64,
    pub value: Value,
}

#[derive(Debug, Clone, Default)]
pub struct StaticIntervalTree {
    intervals: Vec<Interval>,
    tree: Vec<usize>,
    max_ends: Vec<usize>,
    height: usize,
}

impl StaticIntervalTree {
    pub fn new(intervals: impl IntoIterator<Item = Interval>) -> Self {
        let intervals: Vec<_> = intervals.into_iter().collect();
        let size = intervals.len();
        let height = (size + 1).next_power_of_two().ilog2() as usize;
        let mut tree = vec![usize::MAX; (1usize << height).saturating_sub(1)];
        let mut max_ends = vec![usize::MAX; size];
        let mut indices: Vec<_> = (0..size).collect();
        indices.sort_by(|left, right| intervals[*left].start.total_cmp(&intervals[*right].start));

        if size > 0 {
            build(
                &intervals,
                &indices,
                &mut tree,
                &mut max_ends,
                0,
                0,
                size - 1,
            );
        }

        Self {
            intervals,
            tree,
            max_ends,
            height,
        }
    }

    pub fn size(&self) -> usize {
        self.intervals.len()
    }
    pub fn height(&self) -> usize {
        self.height
    }
    pub fn intervals(&self) -> &[Interval] {
        &self.intervals
    }

    pub fn query_point(&self, point: f64) -> Vec<&Interval> {
        self.query_interval(point, point)
    }

    pub fn query_interval(&self, start: f64, end: f64) -> Vec<&Interval> {
        if self.intervals.is_empty() {
            return Vec::new();
        }
        let mut matches = Vec::new();
        let mut stack = vec![0];
        while let Some(node) = stack.pop() {
            let Some(&index) = self.tree.get(node) else {
                continue;
            };
            if index == usize::MAX || start > self.intervals[self.max_ends[index]].end {
                continue;
            }
            let left = node * 2 + 1;
            if self
                .tree
                .get(left)
                .is_some_and(|index| *index != usize::MAX)
            {
                stack.push(left);
            }
            let interval = &self.intervals[index];
            if end >= interval.start && start <= interval.end {
                matches.push(interval);
            }
            if end < interval.start {
                continue;
            }
            let right = node * 2 + 2;
            if self
                .tree
                .get(right)
                .is_some_and(|index| *index != usize::MAX)
            {
                stack.push(right);
            }
        }
        matches
    }
}

fn build(
    intervals: &[Interval],
    indices: &[usize],
    tree: &mut [usize],
    max_ends: &mut [usize],
    node: usize,
    low: usize,
    high: usize,
) -> usize {
    let middle = low + (high - low) / 2;
    let index = indices[middle];
    tree[node] = index;
    let mut max_index = index;
    if low < middle {
        let left = build(
            intervals,
            indices,
            tree,
            max_ends,
            node * 2 + 1,
            low,
            middle - 1,
        );
        if intervals[left].end > intervals[max_index].end {
            max_index = left;
        }
    }
    if middle < high {
        let right = build(
            intervals,
            indices,
            tree,
            max_ends,
            node * 2 + 2,
            middle + 1,
            high,
        );
        if intervals[right].end > intervals[max_index].end {
            max_index = right;
        }
    }
    max_ends[index] = max_index;
    max_index
}
